### DISCUSSION TITLE
ESP Flashing not working via cfcloader (Failed to flash: Failed to update deck bcAI:esp)

### DISCUSSION THREAD

**Discussion by whoenig** [bitcraze/discussions#202] [Q&A] [answered]
I have a CF 2.1 with a single AI Deck 1.1 Rev D attached. I use master/main of STM32 firmware, 2022.05 release of NRF firmware, master/main of cflib and cfclient (as of today).

I followed the instructions to update the AI deck (just a single deck is installed!). Using the cfclient to update, the ESP updates maybe 1 out of 5 tries. Using (from the https://github.com/bitcraze/aideck-esp-firmware)

```
cfloader flash build/aideck_esp.bin deck-bcAI:esp-fw -w radio://0/80/2M
Reset to bootloader mode ...
Could not save cache, no writable directory
Could not save cache, no writable directory
Deck bcAI:esp, reset to bootloader
Failed to flash: Failed to update deck bcAI:esp
```

never works. Updating via JTAG seems to work, although the flashed ESP binary does not show any log output in the cfclient, as expected, so it is difficult to verify that a binary is indeed flashed/executed.

**Comment by knmcguire** [comment]
Hi! 
There has been some changes that requires an latest update of the firmware and esp indeed: https://github.com/bitcraze/aideck-esp-firmware/pull/16. (and 17)

I tried this with the latest master of the STM firmware and cfclient, and flashing the latest ESP, and it seems to work for me with:

`CPX: ESP32: I (831) SYS: Initialized
`
But mine are with verified bootloaders and such so perhaps not the problem here, so let's debug it further:

- With the jtag update, did you put the binary at the right memory? That it didn't flash over the bootloader
- How old is this particular aideck you are having issues with? If you have the original bag, you can see the date there.
- Are you able to see the ESP output if you use the release version of the firmware?

@evoggy is still out for another week so hopefully he'll be able to help out as well once he is back.

**Reply by whoenig** [reply]
Thanks for the fast answer!

1. JTAG Update: Yes, I think so. I simply followed the instructions at https://github.com/bitcraze/aideck-esp-firmware using the provided docker images. This command flashes both the ESP bootloader and the application. This does have a bit lower priority for me, since I'd like to flash over the air, ideally:-)
2. I recently bought 4 and I try to bring them up (adding them to our network that is). I tested with two of them and have the same issue. All 4 were manufactured in 04/16/22 (according to the bag).
3. If the cfclient actually manages to flash the ESP, the output works as expected. None of the two boards were easy to update that way. For the first board it took 5 attempts (each time it also reflashes STM and NRF, so it takes a while...). For the second board I gave up after 3 attempts. In most cases, the cfclient just doesn't update the ESP even without any warning. The console (where cfclient was started from), also doesn't show any indication that/why flashing of the ESP didn't work.

I've seen these baudrate changes, that's why I also tried with the latest STM32 firmware. The key issue of having problems just updating the firmware happens also with the 2022.05 release of the firmware. I have not yet tried downgrading my cflib/cfclient - perhaps this will work better.

**Reply by whoenig** [reply]
Update: when downgrading to cfclient 2022.5 and cflib 0.1.19 the same issue persists: I can't flash via cfloader on the command line, and also not via cfclient (4 attempts). The Crazyradio PA (default firmware) is pretty much next to the Crazyflie and the link quality in the client is 100%.

**Reply by knmcguire** [reply]
Maybe not related, but perhaps also the location of the crazyflie compared to the crazyradio makes a difference. I've placed them about 50 cm from my CR but sometimes when it's closer, it seems to be in a deadzone somehow, which improves moving the crazyflie closer or further away.

It's probably not it but just something I noticed that also makes flashing more difficult.

I'll try an aideck from the same batch soon to see if anything happened with the production firmware

**Reply by knmcguire** [reply]
I'll give you an update tomorrow! I didn't manage to grab a new aideck today.

**Reply by whoenig** [reply]
Thanks! It works with an older AI deck, so I assume it is indeed something related to the newer batch.

**Comment by knmcguire** [comment]
Hi! So just a small update anyway, as had to wait for something so I decided to go ahead and grab an new aideck in the mean time. 

So my process was:

1. Update crazyfie + aideck combo with cfclient to 2022.5
2. Update esp with `python.exe -m cfloader flash build/aideck_esp.bin deck-bcAI:esp-fw -w radio://0/40/2M/E7E7E7E703`
3. update stm firmware with `make cload`

It still works for me and I get `SYS:initialized `in the console. If anything I think it went a bit slower on step 2?

So this will still require some investigation! Seems like something else is going on that we are missing

**Comment by knmcguire** [comment]
I tried some more again today. 

I had a little bit more problems of flashing the AIdeck this morning, but probably because others around me were using a crazyflie as well at that point. But usually after a retry it managed to flash the aideck anyway.

Bootloaders we can perhaps rule out as you did an jtag upload of the bootloader already. 

Another thing that you can try, is to perhaps change the channel and the address the crazyflie and see if that works better? There is something fishy going on the crazyradio's role in this but I can not be sure.

**Comment by whoenig** [comment, marked as answer]
I also tried more yesterday and today and both old and new decks work fine for me on a different Crazyflie. The Crazyflie where it doesn't work has no other known issues (and also no problem detecting/initializing the AI deck), so I am not sure what's going on - perhaps a faulty UART... Perhaps with that I can close it.

**Reply by knmcguire** [reply]
ah that is also a possible cause. Not something that should break easily on a crazyflie but for the aideck it needs to be flawless? Anyway thanks for the update!
